packages/server/api/src/app/workers/redis/redis-rate-limiter.ts

```diff
@@ -0,0 +1,110 @@
+import { AppSystemProp, JobType, logger, OneTimeJobData, system, WebhookJobData } from '@activepieces/server-shared'
+import { assertNotNullOrUndefined, isNil } from '@activepieces/shared'
+import { Job, Queue, Worker } from 'bullmq'
+import dayjs from 'dayjs'
+
+import { createRedisClient, getRedisConnection } from '../../database/redis-connection'
+import { AddParams } from '../queue/queue-manager'
+import { redisQueue } from './redis-queue'
+
+
+let redis = getRedisConnection() 
+
+const RATE_LIMIT = system.getNumberOrThrow(AppSystemProp.RATE_LIMIT)
+
+
+let worker: Worker | null = null
+
+let queue: Queue | null = null
+
+
+
+async function push(projectId: string, jobId: string): Promise<void> {
+    const projectKey = `pending_runs:${projectId}`
+    await redis.sadd(projectKey, jobId)
+    await redis.expire(projectKey, 600)
+}
+
+async function poll(projectId: string): Promise<string | null> {
+    const projectKey = `pending_runs:${projectId}`
+    const jobId = await redis.spop(projectKey)
+    return jobId
+}
+
+export const redisRateLimiter = {
+
+    async init(): Promise<void> {
+        redis = getRedisConnection()
+
+        if (isNil(queue)) {
+            queue = new Queue(
+                'rateLimitJobs',
+                {
+                    connection: createRedisClient(),
+                    defaultJobOptions: {
+                        attempts: 5,
+                        backoff: {
+                            type: 'exponential',
+                            delay: 8 * 60 * 1000,
+                        },
+                        removeOnComplete: true,
+                    },
+                },
+            )
+            await queue.waitUntilReady()
+        }
+        if (isNil(worker)) {
+            worker = new Worker<AddParams<JobType.ONE_TIME | JobType.WEBHOOK>>('rateLimitJobs', async (job) =>{
+                await redisQueue.add(job.data)
+            }, {
+                connection: createRedisClient(),
+                lockDuration: dayjs.duration(30, 'seconds').asMilliseconds(),
+                maxStalledCount: 5,
+                stalledInterval: 30000,
+            })
+            await worker.waitUntilReady()
+        }
+    },
```

Comment:
I would remove the null checks and add the following.

Add this to assertions.ts
```suggestion

export function assertNull<T>(
    value: T | null,
    fieldName: string,
): asserts value is T {
    if (value !== null) {
        throw new Error(`${fieldName} should be null`)
    }
} 
```

and then use it there
```suggestion
    async init(): Promise<void> {
        assertNull(redis, 'redis is not null')
        assertNull(queue, 'queue is not null')
        assertNull(worker, 'worker is not null')
        redis = getRedisConnection()
        queue = new Queue(
            'rateLimitJobs',
            {
                connection: createRedisClient(),
                defaultJobOptions: {
                    attempts: 5,
                    backoff: {
                        type: 'exponential',
                        delay: 8 * 60 * 1000,
                    },
                    removeOnComplete: true,
                },
            },
        )
        await queue.waitUntilReady()
        worker = new Worker<AddParams<JobType.ONE_TIME | JobType.WEBHOOK>>('rateLimitJobs', async (job) => {
            await redisQueue.add(job.data)
        }, {
            connection: createRedisClient(),
            lockDuration: dayjs.duration(30, 'seconds').asMilliseconds(),
            maxStalledCount: 5,
            stalledInterval: 30000,
        })
        await worker.waitUntilReady()
    },
```